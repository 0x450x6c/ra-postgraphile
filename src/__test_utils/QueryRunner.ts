/* eslint-disable @typescript-eslint/no-var-requires */
import { SchemaLink } from 'apollo-link-schema'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { ApolloClient } from 'apollo-client'
require('dotenv').config()
import { Pool } from 'pg'
import { createPostGraphileSchema, withPostGraphileContext } from 'postgraphile'

import PgSimplifyInflectorPlugin from '@graphile-contrib/pg-simplify-inflector'
// @ts-ignore
import PgConnectionFilterPlugin from 'postgraphile-plugin-connection-filter'

export async function makeQueryRunner(
  connectionString = process.env.DATABASE_URL || 'postgres:///',
  schemaName = process.env.DATABASE_SCHEMA || 'app_public',
  options = {
    appendPlugins: [PgSimplifyInflectorPlugin, PgConnectionFilterPlugin],
    graphileBuildOptions: {
      connectionFilterComputedColumns: false,
      connectionFilterRelations: true,
      nestedMutationsSimpleFieldNames: true,
    },
  } // See https://www.graphile.org/postgraphile/usage-schema/ for options
) {
  // Create the PostGraphile schema
  const schema = await createPostGraphileSchema(
    connectionString,
    schemaName,
    options
  )

  const pgPool = new Pool({
    connectionString,
  })

  // @ts-ignore
  const apolloClient: ApolloClient<any> = await withPostGraphileContext(
    { pgPool },
    // @ts-ignore
    (context) => {
      const client = new ApolloClient({
        ssrMode: true,
        cache: new InMemoryCache(),
        link: new SchemaLink({ schema, context }),
      })
      return Promise.resolve(client)
    }
  )

  const release = () => {
    pgPool.end()
  }

  return {
    schema,
    release,
    apolloClient,
  }
}
