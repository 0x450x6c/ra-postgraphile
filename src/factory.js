// @flow

import buildGraphQLProvider from 'ra-data-graphql'
import { defaultQueryValueToInputValueMap } from './defaultValueInputTypeMapping'
import type { ProviderOptions } from './types'
import { buildQuery } from './buildQuery'

export const factory = (client: Object, options: ProviderOptions = {}) => {
  const defaultAppliedOptions = {
    queryValueToInputValueMap: {
      ...defaultQueryValueToInputValueMap,
      ...(options.queryValueToInputValueMap || {})
    }
  }

  return buildGraphQLProvider({
    client,
    buildQuery,
    options: defaultAppliedOptions
  })
}
