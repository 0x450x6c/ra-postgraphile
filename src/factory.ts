import buildGraphQLProvider from 'ra-data-graphql'
import { buildQuery } from './buildQuery'
import { defaultQueryValueToInputValueMap } from './defaultValueInputTypeMapping'
import { ProviderOptions, GraphqlProviderOptions } from './types'

export const factory = (
  client: any,
  options: ProviderOptions = { queryValueToInputValueMap: {} },
  graphqlProviderOptions: GraphqlProviderOptions = {}
) => {
  const defaultAppliedOptions = {
    queryValueToInputValueMap: {
      ...defaultQueryValueToInputValueMap,
      ...(options.queryValueToInputValueMap || {})
    }
  }

  return buildGraphQLProvider({
    ...graphqlProviderOptions,
    client,
    buildQuery,
    options: defaultAppliedOptions
  })
}
