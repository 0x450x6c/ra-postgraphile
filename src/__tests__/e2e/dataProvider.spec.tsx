import expect from 'expect'
import { ApolloClient } from 'apollo-client'
import { introspectionFromSchema, IntrospectionQuery } from 'graphql'
import { convertLegacyDataProvider, DataProvider } from 'ra-core'
import { makeQueryRunner } from '../../__test_utils/QueryRunner'
import { factory } from '../../factory'

let client: ApolloClient<any>
let cleanup: () => void
let introspection: IntrospectionQuery
let dataProvider: DataProvider

beforeAll(async () => {
  const { schema, release, apolloClient } = await makeQueryRunner()
  cleanup = release
  client = apolloClient
  introspection = introspectionFromSchema(schema, { descriptions: true })
  const legacyProvider = await factory(
    client,
    { queryValueToInputValueMap: {} },
    { introspection: introspection.__schema }
  )
  dataProvider = convertLegacyDataProvider(legacyProvider)
})

afterAll(() => {
  cleanup()
})

describe('dataProvider', () => {
  it('should construct from a schema', async () => {
    expect(dataProvider).toBeDefined()
  })
})

describe('Primary keys', () => {
  it('should handle non-id field primary keys with `nodeId`', async () => {
    expect(
      await dataProvider.getList('favoriteBook', {
        sort: { field: 'isbn', order: 'ASC' },
        filter: {},
        pagination: { perPage: 10, page: 1 },
      })
    ).toMatchSnapshot()
  })
})

describe('Query Types', () => {
  describe('should getOne', () => {
    it('should find a book', async () => {
      expect(await dataProvider.getOne('book', { id: 1 })).toMatchSnapshot()
    })
    it('should return null if book was not found', async () => {
      expect(await dataProvider.getOne('book', { id: 5 })).toMatchSnapshot()
    })
  })

  describe('should getList', () => {
    it('should list all books without filter and sorting', async () => {
      expect(
        await dataProvider.getList('book', {
          // @ts-ignore
          sort: {},
          filter: {},
          pagination: { perPage: 10, page: 1 },
        })
      ).toMatchSnapshot()
    })

    it('should list all books with sorting', async () => {
      expect(
        await dataProvider.getList('book', {
          sort: { field: 'id', order: 'ASC' },
          filter: {},
          pagination: { perPage: 10, page: 1 },
        })
      ).toMatchSnapshot()
    })

    it('should filter books', async () => {
      expect(
        await dataProvider.getList('book', {
          sort: { order: 'ASC', field: 'id' },
          filter: { id: 1 },
          pagination: { perPage: 10, page: 1 },
        })
      ).toMatchSnapshot()
    })

    describe('View', () => {
      it('should work with views', async () => {
        expect(
          await dataProvider.getList('allFavoriteBooks', {
            sort: { field: 'id', order: 'ASC' },
            filter: {},
            pagination: { perPage: 10, page: 1 },
          })
        ).toMatchSnapshot()
      })
    })

    describe('Error resilience', () => {
      it('should fail on unknown resources', () => {
        expect(() => dataProvider.getOne('unknownResource', { id: 1 })).toThrow(
          Error('Type "UnknownResource" did not exist in the introspection result.')
        )
      })

      it('should allow lowercase fields to sort on', async () => {
        expect(
          await dataProvider.getList('book', {
            sort: { order: 'asc', field: 'id' },
            filter: {},
            pagination: { perPage: 10, page: 1 },
          })
        ).toMatchSnapshot()
      })

      it('should work on uppercase resource names', async () => {
        expect(await dataProvider.getOne('Book', { id: 1 })).toMatchSnapshot()
      })
    })
  })
})
