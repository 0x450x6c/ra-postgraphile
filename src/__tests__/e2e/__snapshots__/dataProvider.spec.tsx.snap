// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Primary keys should handle non-id field primary keys with \`nodeId\` 1`] = `
Object {
  "data": Array [
    Object {
      "__typename": "FavoriteBook",
      "id": "WyJmYXZvcml0ZV9ib29rcyIsIjEyMzQ1Il0=",
      "isbn": "12345",
      "nodeId": "WyJmYXZvcml0ZV9ib29rcyIsIjEyMzQ1Il0=",
    },
    Object {
      "__typename": "FavoriteBook",
      "id": "WyJmYXZvcml0ZV9ib29rcyIsIjUxMjMxIl0=",
      "isbn": "51231",
      "nodeId": "WyJmYXZvcml0ZV9ib29rcyIsIjUxMjMxIl0=",
    },
  ],
  "total": 2,
}
`;

exports[`Query Types should getList Error resilience should allow lowercase fields to sort on 1`] = `
Object {
  "data": Array [
    Object {
      "__typename": "Book",
      "authorId": 1,
      "id": 1,
      "isbn": "12345",
      "name": "Moby Dick or The Whale",
      "nodeId": "WyJib29rcyIsMV0=",
    },
    Object {
      "__typename": "Book",
      "authorId": 2,
      "id": 2,
      "isbn": "51231",
      "name": "The China Study",
      "nodeId": "WyJib29rcyIsMl0=",
    },
    Object {
      "__typename": "Book",
      "authorId": 3,
      "id": 3,
      "isbn": "3221123",
      "name": "How Not to Die",
      "nodeId": "WyJib29rcyIsM10=",
    },
  ],
  "total": 3,
}
`;

exports[`Query Types should getList should filter books 1`] = `
Object {
  "data": Array [
    Object {
      "__typename": "Book",
      "authorId": 1,
      "id": 1,
      "isbn": "12345",
      "name": "Moby Dick or The Whale",
      "nodeId": "WyJib29rcyIsMV0=",
    },
  ],
  "total": 1,
}
`;

exports[`Query Types should getList should list all books with sorting 1`] = `
Object {
  "data": Array [
    Object {
      "__typename": "Book",
      "authorId": 1,
      "id": 1,
      "isbn": "12345",
      "name": "Moby Dick or The Whale",
      "nodeId": "WyJib29rcyIsMV0=",
    },
    Object {
      "__typename": "Book",
      "authorId": 2,
      "id": 2,
      "isbn": "51231",
      "name": "The China Study",
      "nodeId": "WyJib29rcyIsMl0=",
    },
    Object {
      "__typename": "Book",
      "authorId": 3,
      "id": 3,
      "isbn": "3221123",
      "name": "How Not to Die",
      "nodeId": "WyJib29rcyIsM10=",
    },
  ],
  "total": 3,
}
`;

exports[`Query Types should getList should list all books without filter and sorting 1`] = `
Object {
  "data": Array [
    Object {
      "__typename": "Book",
      "authorId": 1,
      "id": 1,
      "isbn": "12345",
      "name": "Moby Dick or The Whale",
      "nodeId": "WyJib29rcyIsMV0=",
    },
    Object {
      "__typename": "Book",
      "authorId": 2,
      "id": 2,
      "isbn": "51231",
      "name": "The China Study",
      "nodeId": "WyJib29rcyIsMl0=",
    },
    Object {
      "__typename": "Book",
      "authorId": 3,
      "id": 3,
      "isbn": "3221123",
      "name": "How Not to Die",
      "nodeId": "WyJib29rcyIsM10=",
    },
  ],
  "total": 3,
}
`;

exports[`Query Types should getOne should find a book 1`] = `
Object {
  "data": Object {
    "__typename": "Book",
    "authorId": 1,
    "id": 1,
    "isbn": "12345",
    "name": "Moby Dick or The Whale",
    "nodeId": "WyJib29rcyIsMV0=",
  },
}
`;

exports[`Query Types should getOne should return null if book was not found 1`] = `
Object {
  "data": null,
}
`;
